Training/Training/Training.c

```diff
@@ -1,14 +1,194 @@
+/// Function which checks if a given string is palindrome or not
+/// </summary>
+int PalindromeChecker (char* phrase) {
+   int i = 0, j = 0, length, newLength, flag = 0;
+   char temp[MAX];
+   length = (int)strlen (phrase);
+   if (length > MAX - 2) {
+      printf ("\nLimit exceeded");
+      return INVALID;
+   }
+   //Eliminating special characters by moving the remaining character into a temporary array
+   while (i < length) {
+      if (isalnum (phrase[i]) != 0)
+         temp[j++] = phrase[i];
+      i++;
+   }
+   temp[j] = '\0';
+   newLength = (int)strlen (temp);
+   //Checking invalid text
+   if (newLength == 0)
+      return INVALID;
+   i = 0;
+   while (i < newLength / 2) {
+      if (toupper (temp[i]) == toupper (temp[newLength - 1 - i]))
+         flag++;
+      i++;
+   }
+   if (flag == newLength / 2)
+      return PALINDROME;
+   else
+      return NOTPALINDROME;
+}
+
+/// <summary>
+/// Function which checks if the input number is palindrome or not
+/// </summary>
+int NumberReverser (int number) {
+   int rem, numTemp;
+   long long int rev = 0;
+   numTemp = number;
+   while (numTemp != 0) {
+      rem = numTemp % 10;
+      rev = rev * 10 + rem;
+      numTemp /= 10;
+   }
+   printf ("\nReversed: %lld", rev);
+   if (number >= 0 && number == rev)
+      return PALINDROME;
+   else
+      return NOTPALINDROME;
+}
+
+/// <summary>
+/// Function to run test cases
+/// </summary>
+void TestCases (int* numInputs, int numLength, char* strInputs[], int strLength, int* numExpected, int* strExpected) {
+   int* strOutputs = NULL, * numOutputs = NULL;
+   do {
+      strOutputs = (int*)malloc (strLength * sizeof (int));
+   } while (strOutputs == NULL);
+   for (int i = 0; i < strLength; i++) {
+      printf ("\nInput: %s", strInputs[i]);
+      strOutputs[i] = PalindromeChecker (strInputs[i]);
+      printf ("\nExpected: %d", strExpected[i]);
+      printf ("\nOutput: %d", strOutputs[i]);
+      if (strOutputs[i] == strExpected[i])
+         printf (ANSI_COLOR_GREEN"\nPassed\n"ANSI_COLOR_RESET);
+      else
+         printf (ANSI_COLOR_RED"\nFailed\n"ANSI_COLOR_RESET);
+   }
+   do {
+      numOutputs = (int*)malloc (numLength * sizeof (int));
+   } while (numOutputs == NULL);
+   for (int i = 0; i < numLength; i++) {
+      printf ("\nInput: %d", numInputs[i]);
+      numOutputs[i] = NumberReverser (numInputs[i]);
+      printf ("\nExpected: %d", numExpected[i]);
+      printf ("\nOutput: %d", numOutputs[i]);
+      if (numOutputs[i] == numExpected[i])
+         printf (ANSI_COLOR_GREEN"\nPassed\n"ANSI_COLOR_RESET);
+      else
+         printf (ANSI_COLOR_RED"\nFailed\n"ANSI_COLOR_RESET);
+   }
+}
+
+int main () {
+   char ch = '\n', choice[3], * result, term;
+   while (1) {
+      printf ("Menu");
+      printf ("\n~~~~");
+      printf ("\n1. Palindrome Checker");
+      printf ("\n2. Number Reverser");
+      printf ("\n3. Run Test Cases");
+      printf ("\n\nEnter your choice: ");
+      fgets (choice, 3, stdin);

```

Comment:
Can write all the statements in a single printf statement. Reducing the number of calls to the system can speed up your program. Apply this change wherever possible.